Small fixes

SergeantChris opened the pull request:
This PR includes small changes and fixes I had to implement in order to work smoothly with a large number of models.
These are the following:
* `netsetbase.py` (and `feature_extraction.py`): Existing code in the case `elif network_layers` gets a list of layers from manually predefined json files (which are the top level layer blocks of known models). If the json does not exist then all layers are used. 
Instead, here I propose an alternative that worked well for me to generalize getting the top level layer blocks of any model (not restricted to predefined libraries). 
To still support the option of extracting all layers, I extended the user parameter `layers_to_extract` with string options `"all"` and `"top_level"`, and also maintained the option of the json with a third option `"json"`.
* `rsa.py` (and `distance_functions.py`, `noiseceiling.py` in point 2): 
(1) Added the option `layer_skips` in the RSA object, which allows the user to pass a list of layer names for which RSA should not be run. Can be useful to run RSA on different subsets of extracted layer features. Alternatively the user could run RSA on all layers and only after filter out the dataframe.
(2) Added the option `squared` in the RSA object. This gives the user the choice on whether to compute the average of squared correlation across subjects, or simply the average correlation across subjects. With the existing code so far, there is only the option to get the root afterwards (in the plotting), but this is problematic as (A) the root of the mean of squared correlations is not equal to the mean of correlations (they differ by the variance factor) and (B) negative correlations are ignored and cannot be recovered.
(3) Instead of deleting the correlation array, it is now included in the dataframe, as there are many cases that this might be needed by the user (e.g. custom plotting). 
* `feature_extraction.py`, `feature_iterator.py`, `dim_reduction.py`: There were some edge cases in model layers that had multiple dimensions that needed squeezing, or were scalar. This small bug required these minor fixes.

TODO: Update user-guide notebooks with new parameters.

ToastyDom commented:
Hey thank you so much for your PR!
I checked the functions and they work really well! I especially like the layer_to_extract improvement! And thank you for pointing out the issue with the root during the plotting!

The only issue I saw is that within the github-tests we now have to adapt the ground truth for the evaluation metric because we now have a dataframe shape mismatch with the added column.

There is another new error with the custom_cosine function in the tests but I could not tell right now where this issue might come from. I will look into that and a better way to deal with the HTTP errors which I thought I fixed...

But beyond the dataframe shape mismatch and and the update of user guide books you can go ahead and merge! :)

Thank you so much!

SergeantChris commented:
Thanks for reviewing @ToastyDom !
I fixed the test-cases with the shape mismatch for the added column, and updated the notebook documentation.

I also could not find the issue in the distance function test case (looks like a tensor device error?)

You can check the changes in the notebooks through the ReviewNB link (https://app.reviewnb.com/cvai-roig-lab/Net2Brain/pull/91/).
